File: Python/Dynamic Programming/139. Word Break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        wordDict = set(wordDict)
        n = len(s)
        dp = [False for _ in range(n+1)]
        dp[0] = True
        for i in range(1, n+1):
            for j in range(max(0, i-20), i):
                tmp = dp[j] and s[j:i] in wordDict
                if tmp: 
                    dp[i] = True
                    break
        return dp[n]
            

File: C++/Dynamic Programming/139. Word Break.cpp
class Solution {
public:
    
    bool helper(string& s, unordered_set<string>& words, int start, unordered_map<int, bool>& dp) {
        if (start == s.size()) {
            return true;
        }
        if (dp.count(start)) {
            return dp[start];
        }
        for (int len = 1; len <= s.size() - start; len++) {
            if (words.count(s.substr(start, len))) {
                if (helper(s, words, start+len, dp)) {
                    dp[start] = true;
                    return true;
                }
            }
        }
        dp[start] = false;
        return false;
    }
    
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> words(wordDict.begin(), wordDict.end());
        unordered_map<int, bool> dp;
        return helper(s, words, 0, dp);
    }

};